infra/lambda/datasets-endpoint/function.py

```diff
@@ -0,0 +1,482 @@
+"""
+Dataset Lambda handler function.
+"""
+
+import uuid
+from datetime import datetime
+
+import boto3
+from dateutil.tz import gettz
+from jsonschema import ValidationError, validate
+
+DYNAMODB = boto3.client("dynamodb")
+
+DS_PRIMARY_KEYS = ("id", "type")
+DS_ATTRIBUTES = ("title", "owning_group")
+DS_ATTRIBUTES_EXT = ("created_at",)
+
+REQUEST_SCHEMA = {
+    "type": "object",
+    "properties": {
+        "httpMethod": {"type": "string", "enum": ["GET", "POST", "PATCH", "DELETE"]},
+        "body": {"type": "object"},
+    },
+    "required": ["httpMethod", "body"],
+}
+
+# TODO: implement GET response paging
+# TODO: allow Dataset delete only if no Dataset Version exists
+# TODO: don't assume that all Dataset attributes are strings ("S")
+
+
+# fmt: off
+def lambda_handler(event, context):  # pylint:disable=unused-argument,inconsistent-return-statements,too-many-return-statements
+    """Main Lambda entry point."""
+# fmt: on
+
+    # request validation
+    try:
+        validate(event, REQUEST_SCHEMA)
+        method = event["httpMethod"]
+    except ValidationError as err:
+        return {"statusCode": 400, "body": {"message": f"Bad Request: {err.message}."}}
+
+    if method == "POST":
+        return post_method(event)
+
+    if method == "GET" and "id" in event["body"] and "type" in event["body"]:
+        return get_method_single(event)
+
+    if method == "GET" and ("title" in event["body"] or "owning_group" in event["body"]):
+        return get_method_filter(event)
+
+    if method == "GET" and event["body"] == {}:
+        return get_method_all()
+
+    if method == "PATCH":
+        return patch_method(event)
+
+    if method == "DELETE":
+        return delete_method(event)
+
+
+def post_method(payload):
+    """POST: Create Dataset."""
+
+    BODY_SCHEMA = {
+        "type": "object",
+        "properties": {
+            "type": {
+                "type": "string",
+                "enum": ["IMAGE", "RASTER"],
+            },
+            "title": {"type": "string"},
+            "owning_group": {"type": "string"},
+        },
+        "required": ["type", "title", "owning_group"],
+    }
+
+    # request body validation
+    try:
+        req_body = payload["body"]
+        validate(req_body, BODY_SCHEMA)
+    except ValidationError as err:
+        return {"statusCode": 400, "body": {"message": f"Bad Request: {err.message}."}}
+
+    # get PKs
+    pk = {}
+    pk["id"] = uuid.uuid4().hex[:6]
+    pk["type"] = payload["body"]["type"]
+
+    # get attributes
+    attr = {}
+    for a in DS_ATTRIBUTES:
+        attr[a] = payload["body"][a]
+
+    curr_time = datetime.utcnow()
+    attr["created_at"] = str(curr_time.replace(tzinfo=gettz("UTC")))
+
+    # make sure that requested type/title doesn't already exist in DB
```

Comment:
> Our checking code usually also duplicates some of the code in the underlying system (DynamoDB), which is redundant since the system will perform those checks when attempting to create an instance.

I don't think this is true. There is no way I know how to enforce some `UNIQUE` like constraint on non primary key attribute in DynoDB. That's why it needs to be checked on application level. `query` on indexed attribute is fast.

> No matter how thorough the checking code is, it cannot even theoretically guarantee that the instance will actually be created. So now we have to handle errors from two places in this process.

That's true but for now I don't have any better idea.

Reply:
[Transactions](https://aws.amazon.com/blogs/database/simulating-amazon-dynamodb-unique-constraints-using-transactions/) seems to be one solution, but since this is more complex than in an RDBMS and we won't have too much traffic to begin with we can just leave it for now. Thanks for the explanation!